src/core/lib/event_engine/windows/socket.h

```diff
@@ -0,0 +1,115 @@
+
+class WinSocket final : public SocketNotifier {
+ public:
+  // State related to a Read or Write socket operation
+  class OpState {
+   public:
+    explicit OpState(WinSocket* win_socket) noexcept;
+    // Signal a result has returned
+    // If a callback is already primed for notification, it will be executed via
+    // the WinSocket's EventEngine. Otherwise, a "pending iocp" flag will
+    // be set.
+    void SetReady();
+    // Set error results for a completed op
+    void SetError(int wsa_error);
+    // Retrieve results of overlapped operation (via Winsock API)
+    void GetOverlappedResult();
+    // OVERLAPPED, needed for Winsock API calls
+    OVERLAPPED* overlapped() { return &overlapped_; }
+    // Data from the previous operation, set via GetOverlappedResult
+    DWORD bytes_transferred() const { return bytes_transferred_; }
+    // Previous error if set.
+    int wsa_error() const { return wsa_error_; }
+    EventEngine::Closure* closure() { return closure_; }
+
+   private:
+    friend class WinSocket;
+
+    OVERLAPPED overlapped_;
+    WinSocket* win_socket_ = nullptr;
+    EventEngine::Closure* closure_ = nullptr;
+    bool has_pending_iocp_ = false;
+    DWORD bytes_transferred_;
+    int wsa_error_;
+  };
+
+  WinSocket(SOCKET socket, EventEngine* event_engine) noexcept;
+  ~WinSocket();
+  // Calling NotifyOnRead means either of two things:
+  //  - The IOCP already completed in the background, and we need to call
+  //    the callback now.
+  //  - The IOCP hasn't completed yet, and we're queuing it for later.
+  void NotifyOnRead(EventEngine::Closure* on_read) override;
+  void NotifyOnWrite(EventEngine::Closure* on_write) override;
+  void SetReadable() override;
+  void SetWritable() override;
+  // Schedule a shutdown of the socket operations. Will call the pending
+  // operations to abort them. We need to do that this way because of the
+  // various callsites of that function, which happens to be in various
+  // mutex hold states, and that'd be unsafe to call them directly.
+  void MaybeShutdown(absl::Status why) override;
+  bool IsShutdown() override;
+
+  // Return the appropriate OpState for a given OVERLAPPED
+  // Returns nullptr if the overlapped does not match either read or write ops.
+  OpState* GetOpInfoForOverlapped(OVERLAPPED* overlapped);
+  // -------------------------------------------------
+  // TODO(hork): We need access to these for WSA* ops in TCP code.
+  // Maybe we can encapsulate these calls inside of the OpState class. Would need
+  // to rename it.
+  OpState* read_info() { return &read_info_; }
+  OpState* write_info() { return &write_info_; }
+  // -------------------------------------------------
+  // Accessor method for underlying socket
+  SOCKET socket();
+
+ private:
+  void NotifyOnReady(OpState& info, EventEngine::Closure* callback);
+
+  SOCKET socket_;
+  grpc_core::Mutex mu_;
+  bool is_shutdown_ ABSL_GUARDED_BY(mu_) = false;
+  EventEngine* event_engine_;
```

Comment:
Is there a more restrictive interface we could offer here?

Reply:
Yes, as we discussed, I was considering a subset of the EventEngine API, something like 

```
class Executor {
  virtual ~Executor() = default;
  virtual void Run(Closure* closure) = 0;
  virtual void Run(absl::AnyInvocable<void()> closure) = 0;  
};
```

The idea would be, though, that some EventEngine implementation would be plugged into it. i.e., `class MyEventEngine : public EventEngine, Executor { ... };`. So it doesn't make much difference until there are alternative implementations to plug into it.

Reply:
I take that back, I'll extract the common Executor functionality and keep it separate, and have the engines take ownership of one.

Reply:
Done.